File: HackerRank/Algorithms/Implementation/Equal Stacks/es.rb
#!/bin/ruby

n1,n2,n3 = gets.strip.split(' ')
n1 = n1.to_i
n2 = n2.to_i
n3 = n3.to_i
h1 = gets.strip
h1 = h1.split(' ').map(&:to_i)
h2 = gets.strip
h2 = h2.split(' ').map(&:to_i)
h3 = gets.strip
h3 = h3.split(' ').map(&:to_i)
H1 = h1.reduce :+
H2 = h2.reduce :+
H3 = h3.reduce :+
while H1 != H2 || H2 != H3 do
    while H1 > H2 || H1 > H3 do
        h = h1.shift
        H1 -= h
    end
    
    while H2 > H1 || H2 > H3 do
        h = h2.shift
        H2 -= h
    end
    
    while H3 > H1 || H3 > H2 do
        h = h3.shift
        H3 -= h
    end
end

puts H1

File: HackerRank/Algorithms/Implementation/Equal Stacks/es.py
#!/bin/python3

import sys


n1,n2,n3 = input().strip().split(' ')
n1,n2,n3 = [int(n1),int(n2),int(n3)]
h1 = [int(h1_temp) for h1_temp in input().strip().split(' ')]
h2 = [int(h2_temp) for h2_temp in input().strip().split(' ')]
h3 = [int(h3_temp) for h3_temp in input().strip().split(' ')]

H1 = sum(h1)
H2 = sum(h2)
H3 = sum(h3)

while not (H1 == H2 == H3):
    while H1 > H2 or H1 > H3:
        h = h1.pop(0)
        H1 -= h
    while H2 > H1 or H2 > H3:
        h = h2.pop(0)
        H2 -= h
    while H3 > H1 or H3 > H2:
        h= h3.pop(0)
        H3 -= h
print(H1)